src/writers/sink.creator.cpp

```diff
@@ -0,0 +1,409 @@
+}
+
+bool
+zarr::SinkCreator::create_shard_sinks(
+  const std::string& base_uri,
+  const std::vector<Dimension>& dimensions,
+  std::vector<std::shared_ptr<Sink>>& shard_sinks)
+{
+    return make_part_sinks_(
+      base_uri, dimensions, common::shards_along_dimension, shard_sinks);
+}
+
+bool
+zarr::SinkCreator::create_v2_metadata_sinks(
+  const std::string& base_uri,
+  size_t n_arrays,
+  std::vector<std::shared_ptr<Sink>>& metadata_sinks)
+{
+    std::queue<std::string> dir_paths, file_paths;
+
+    file_paths.emplace(".metadata"); // base metadata
+    file_paths.emplace("0/.zattrs"); // external metadata
+    file_paths.emplace(".zattrs");   // group metadata
+
+    for (auto i = 0; i < n_arrays; ++i) {
+        dir_paths.push(std::to_string(i));
+        file_paths.push(std::to_string(i) + "/.zarray"); // array metadata
+    }
+
+    return make_metadata_sinks_(
+      base_uri, dir_paths, file_paths, metadata_sinks);
+}
+
+bool
+zarr::SinkCreator::create_v3_metadata_sinks(
+  const std::string& base_uri,
+  size_t n_arrays,
+  std::vector<std::shared_ptr<Sink>>& metadata_sinks)
+{
+    std::queue<std::string> dir_paths, file_paths;
+
+    dir_paths.emplace("meta");
+    dir_paths.emplace("meta/root");
+
+    file_paths.emplace("zarr.json");
+    file_paths.emplace("meta/root.group.json");
+    for (auto i = 0; i < n_arrays; ++i) {
+        file_paths.push("meta/root/" + std::to_string(i) + ".array.json");
+    }
+
+    return make_metadata_sinks_(
+      base_uri, dir_paths, file_paths, metadata_sinks);
+}
+
+bool
+zarr::SinkCreator::make_dirs_(std::queue<std::string>& dir_paths)
+{
+    if (dir_paths.empty()) {
+        return true;
+    }
+
+    std::atomic<bool> all_successful = true;
+
+    const auto n_dirs = dir_paths.size();
+    std::latch latch(n_dirs);
+
+    for (auto i = 0; i < n_dirs; ++i) {
+        const auto dirname = dir_paths.front();
+        dir_paths.pop();
+
+        thread_pool_->push_to_job_queue(
+          [dirname, &latch, &all_successful](std::string& err) -> bool {
+              bool success = false;
+
+              try {
+                  if (fs::exists(dirname)) {
+                      EXPECT(fs::is_directory(dirname),
+                             "'%s' exists but is not a directory",
+                             dirname.c_str());
+                  } else if (all_successful) {
+                      std::error_code ec;
+                      EXPECT(fs::create_directories(dirname, ec),
+                             "%s",
+                             ec.message().c_str());
+                  }
+                  success = true;
+              } catch (const std::exception& exc) {
+                  char buf[128];
+                  snprintf(buf,
+                           sizeof(buf),
+                           "Failed to create directory '%s': %s.",
+                           dirname.c_str(),
+                           exc.what());
+                  err = buf;
+              } catch (...) {
```

Comment:
Every time we use `catch (...)` in this file there's a `catch (const std::exception&)` above it.

Reply:
That's fine, I guess my concern is that we would probably want an unknown exception to break? In the following case, the exception that lands in (...) may be very specific, but we lose the information and instead log `Exception: (unknown)`. This will make it very difficult to understand the logs.

```cpp
} catch (const std::exception& exc) {
    LOGE("Exception: %s\n", exc.what());
} catch (...) {
   LOGE("Exception: (unknown)"); // which exception was thrown? what was the original message? etc
}
```